File: backoffice/src/shared/infrastructure/events/JsonDomainEventDeserializer.py

"""
 *
 * Libraries 
 *
"""

import json
from typing                   import Any
from src.shared.domain        import DomainEvent
from injector                 import inject
from .DomainEventsInformation import DomainEventsInformation
from .DomainEventInformation  import DomainEventInformation

"""
 *
 * Class
 *
"""

class JsonDomainEventDeserializer:

    """
     *
     * Attributes 
     *
    """

    __eventsInformation : DomainEventsInformation

    """
     *
     * Methods 
     *
    """

    @inject
    def __init__( self, eventsInformation : DomainEventsInformation ) -> object:
        self.__eventsInformation = eventsInformation

    def deserialize( self, messageBody : str ) -> DomainEvent:
        # Variables
        messageData      : dict[str, Any]
        eventClass       : type
        eventName        : str
        domainEvent      : DomainEvent
        eventInformation : DomainEventInformation
        # Code
        messageData      = json.loads( messageBody )
        eventName        = messageData['name']
        eventInformation = self.__eventsInformation.findByEventName( eventName )
        eventClass       = eventInformation.getEventClass()
        domainEvent      = eventClass()
        domainEvent = domainEvent.fromPrimitives(
            eventId   = messageData['id'],
            timestamp = messageData['timestamp'],
            data      = messageData['data']
        )
        return domainEvent

File: preparation/src/shared/infrastructure/events/JsonDomainEventDeserializer.cs
/* 
 * 
 * Libraries
 *
*/

using Newtonsoft.Json;
using preparation.src.shared.domain;

/* 
 * 
 * Class
 *
*/

namespace preparation.src.shared.infrastructure {

    public class JsonDomainEventDeserializer { 

        /* 
         * 
         * Attributes
         *
        */

        private DomainEventsInformation eventsInformation;

        /* 
         * 
         * Methods
         *
        */

        public JsonDomainEventDeserializer( DomainEventsInformation eventsInformation ) {
            this.eventsInformation = eventsInformation;
        }

        public DomainEvent deserialize( string messageBody ) {
            // Variables
            Dictionary<string, object> messageData;
            Type                       eventClass;
            string                     eventName;
            DomainEvent                domainEvent;
            DomainEventInformation     eventInformation;
            Dictionary<string, object> data;
            // Code
            messageData      = JsonConvert.DeserializeObject<Dictionary<string, object>>( messageBody );
            eventName        = messageData["name"].ToString();
            eventInformation = eventsInformation.findByEventName( eventName );
            eventClass       = eventInformation.getEventClass();
            domainEvent      = ( DomainEvent )Activator.CreateInstance( eventClass );
            data             = JsonConvert.DeserializeObject<Dictionary<string, object>>( messageData["data"].ToString() );
            domainEvent = ( DomainEvent )domainEvent.fromPrimitives( 
                eventId   : messageData["id"].ToString(), 
                timestamp : Int32.Parse( messageData["timestamp"].ToString() ),
                data      : data
            );
            return domainEvent;
        }
    
    }

}